sdk/go/pulumi/context.go

```diff
@@ -225,9 +225,24 @@ func (ctx *Context) Close() error {
 	return nil
 }
 
-// wait waits for all asynchronous work associated with this context to drain. RPCs may not be queued once wait
+// wait waits for all asynchronous work associated with this context to drain.
+func (ctx *Context) wait() {
+	// Wait for async work to flush.
+	ctx.state.join.Wait()
+
+	// Ensure all outstanding RPCs have completed before proceeding.
+	ctx.state.rpcsLock.Lock()
+	defer ctx.state.rpcsLock.Unlock()
+
+	// Wait until the RPC count hits zero.
+	for ctx.state.rpcs > 0 {
+		ctx.state.rpcsDone.Wait()
+	}
+}
+
+// close waits for all asynchronous work associated with this context to drain. RPCs may not be queued once close
 // returns.
-func (ctx *Context) wait() error {
+func (ctx *Context) close() error {
 	// Wait for async work to flush.
```

Comment:
The first part of this method can just be replaced with `ctx.wait()`

Reply:
That was my first instinct too but I think it preferrable to acquire the lock once. The alternatives were clunky.

Reply:
Fair